source/Ship.cpp

```diff
@@ -4791,3 +5159,375 @@ double Ship::CalculateDeterrence() const
+			continue;
+		other->minWeaponRange = min(other->minWeaponRange, weapon->Range());
+		other->maxWeaponRange = max(other->maxWeaponRange, weapon->Range());
+	}
+	if(other->maxWeaponRange == 0.)
+		other->minWeaponRange = 0.;
+}
+
+
+
+// Perform calculations across all escorts to set private variables
+// escortsHaveOneJump, isEnemyInEscortSystem, isEscortsFullOfFuel, and
+// refuelMissionNpcEscort.  Warning this is an expensive calculation if done too
+// often.  This function is called by AI.cpp while in space and PlayerInfo.cpp
+// while in an outfitter roughly 3 times a second.
+void Ship::UpdateEscortsState()
+{
+	vector<shared_ptr<Ship>> parents;
+	std::shared_ptr<Ship> flagship = GetParent();
+	if(flagship)
+	{
+		parents.emplace_back(flagship);
+		string errmsg = "Error: the following ships form a circular parent relationship: ";
+		while(flagship->GetParent())
+		{
+			auto nextparent = find(parents.begin(), parents.end(), flagship->GetParent());
+			if(nextparent != parents.end())
+			{
+				for(auto next : parents)
+					errmsg += next->name + ", ";
+				errmsg += flagship->GetParent()->name;
+				Logger::LogError(errmsg);
+				break;
+			}
+			flagship = flagship->GetParent();
+			parents.emplace_back(flagship);
+		}
+	}
+	if(!flagship)
+		flagship = shared_from_this();
+	parents.clear();
+
+	if(!flagship)
+		return;
+
+	// Player escorts should only be updated if this ship is owned by the
+	// player.
+	if(flagship->IsYours() && !IsYours())
+		return;
+
+	// Reset the carried ship fuel so that it can be recalculated.
+	flagship->maxCarriedShipFuel = 0.;
+
+	const vector<weak_ptr<Ship>> allEscorts = flagship->GetEscorts();
+	UpdateEscortsState(flagship, allEscorts);
+
+	// Check flagship against itself
+	if(!flagship->IsTankerCarrier() && flagship->IsFuelLow())
+		flagship->isEscortsFullOfFuel = false;
+	shared_ptr<const Ship> flagshipTarget = flagship->GetTargetShip();
+	const Government *gov = flagship->GetGovernment();
+	if(flagshipTarget)
+		if(gov->IsEnemy(flagshipTarget->GetGovernment()) && !flagshipTarget->IsDisabled() && flagshipTarget->IsTargetable())
+			flagship->isEnemyInEscortSystem = true;
+	if(flagship->JumpFuelMissing())
+		flagship->escortsHaveOneJump = false;
+
+	// Set the state for all fighters and drones.
+	for(const weak_ptr<Ship> &ptr : allEscorts)
+	{
+		shared_ptr<Ship> escort = ptr.lock();
+		// Skip this ship as it will be updated at the end.
+		if(escort == shared_from_this())
+			continue;
+		if(!IsEscortedBy(flagship, escort))
+			continue;
+		flagship->UpdateEscortsState(escort);
+		if(!escort->CanBeCarried())
+		{
+			// This covers boarded escorts not in the system.
+			for(const weak_ptr<Ship> &ptr2 : escort->GetEscorts())
+			{
+				shared_ptr<Ship> carry = ptr2.lock();
+				if(!carry)
+					continue;
+				if(!carry->CanBeCarried())
+					continue;
+				flagship->UpdateEscortsState(carry);
+			}
+		}
+	}
+
+	// Update this ship based.  If this is the flagship, then it will adjust weapon stats.
```

Comment:
> this ship based

What does that mean? Could you clarify?
```suggestion
	// Update this ship based. If this is the flagship, then it will adjust weapon stats.
```